pkg/storegateway/indexheader/stream_binary_reader.go

```diff
@@ -388,3 +347,138 @@ func (r *StreamBinaryReader) Close() error {
+	sparseHeader.Symbols = r.symbols.NewSparseSymbol()
+	sparseHeader.PostingsOffsetTable = r.postingsOffsetTable.NewSparsePostingOffsetTable()
+	return sparseHeader
+}
+
+type bucketAndDiskSparseHeaderLoader struct {
+	ctx       context.Context
+	bkt       objstore.Bucket
+	localPath string
+	logger    log.Logger
+}
+
+func (l bucketAndDiskSparseHeaderLoader) SparseHeader() (*indexheaderpb.Sparse, error) {
+	defer func(start time.Time) {
+		level.Debug(l.logger).Log("msg", "read and parsed sparse index header", "path", l.localPath, "elapsed", time.Since(start))
+	}(time.Now())
+
+	sparseData, existsLocally, err := l.sparseHeaderBytes()
+	if err != nil {
+		return nil, err
+	}
+	if len(sparseData) > 0 && !existsLocally {
+		err = l.persistHeaderToDisk(sparseData)
+		if err != nil {
+			level.Error(l.logger).Log("msg", "couldn't persist sparse index to disk; will download from object store next time too", "path", l.localPath, "err", err)
+			// otherwise ignore the error because loading the header is the priority; not persisting it to disk
+		}
+	}
+
+	sparseHeaders, err := l.parseSparseHeader(sparseData)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse sparse index-header: %w", err)
+	}
+	return sparseHeaders, nil
+}
+
+func (l bucketAndDiskSparseHeaderLoader) sparseHeaderBytes() (b []byte, existsLocally bool, _ error) {
+	b, err := os.ReadFile(l.localPath)
+	if err == nil {
+		return b, true, nil
+	}
+	if !os.IsNotExist(err) {
+		level.Warn(l.logger).Log("msg", "failed to read sparse index-header from disk; trying to find it in object store", "err", err)
+	}
+
+	reader, err := l.bkt.Get(l.ctx, block.SparseIndexHeaderFilename)
+	if err != nil {
+		level.Warn(l.logger).Log("msg", "failed to find sparse index-header in object store; recreating", "err", err)
+		return nil, false, err
+	}
+	b, err = io.ReadAll(reader)
+	if err != nil {
+		level.Warn(l.logger).Log("msg", "failed to read sparse index-header from object store; recreating", "err", err)
+		return nil, false, err
+	}
+	return b, false, nil
+}
+
+func (l bucketAndDiskSparseHeaderLoader) parseSparseHeader(sparseData []byte) (*indexheaderpb.Sparse, error) {
+	gzipped := bytes.NewReader(sparseData)
+	gzipReader, err := gzip.NewReader(gzipped)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create sparse index-header reader: %w", err)
+	}
+
+	sparseData, err = io.ReadAll(gzipReader)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read sparse index-header: %w", err)
+	}
+
+	sparseHeaders := &indexheaderpb.Sparse{}
+	if err := sparseHeaders.Unmarshal(sparseData); err != nil {
+		return nil, fmt.Errorf("failed to decode sparse index-header file: %w", err)
+	}
+	return sparseHeaders, nil
+}
+
+// PersistSparseHeader uses protocol buffer to write StreamBinaryReader to disk at path and to object storage at id.String/block.SparseIndexHeaderFilename
+func (l bucketAndDiskSparseHeaderLoader) PersistSparseHeader(sparseHeader *indexheaderpb.Sparse) (err error) {
```

Comment:
Given we (rightly) won't return an error to a user if persisting the sparse header fails, should we add a metric for the number of successful and failed persistence attempts? Then we could alert on failures, which might otherwise go unnoticed for quite a while.